src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java

```diff
@@ -58,29 +58,19 @@ private static boolean canUseLeg(Leg leg) {
     }
 
     /**
-     * Get the expected leg by comparing the current time against the start and end time of each leg.
+     * Get the expected leg by comparing the current position against all points on a leg. Comparing against all points
+     * on a leg gives greater accuracy than just comparing against the start and end of a leg.
      */
     @Nullable
-    public static Leg getExpectedLeg(Instant timeNow, Itinerary itinerary) {
+    public static Leg getExpectedLeg(Coordinates position, Itinerary itinerary) {
         if (canUseTripLegs(itinerary)) {
             for (int i = 0; i < itinerary.legs.size(); i++) {
                 Leg leg = itinerary.legs.get(i);
-                if (i == 0 &&
-                    leg.mode.equalsIgnoreCase("walk") &&
-                    leg.startTime != null &&
-                    timeNow.isBefore(leg.startTime.toInstant())) {
-                    // If the first leg is a walk leg and the traveler has decided to start the trip early.
-                    return leg;
-                }
-                if (leg.startTime != null &&
-                    leg.endTime != null &&
-                    isTimeInRange(
-                        leg.startTime.toInstant(),
-                        // Offset the end time by a faction to avoid exact times being attributed to the wrong leg.
-                        leg.endTime.toInstant().minus(1, ChronoUnit.MILLIS),
-                        timeNow
-                    )) {
-                    return leg;
+                List<Coordinates> allPositions = getAllLegPositions(leg);
+                for (int j = 0; j < allPositions.size() - 1; j++) {
+                    if (isPointBetween(allPositions.get(j), allPositions.get(j + 1), position)) {
```

Comment:
I think this condition is too restrictive. For the initial step, this will never work. For all intermediate steps, this takes over the on-track radii defined as parameters. I thought you had code to snap to a nearest leg segment, could we use that instead?

Reply:
I agree. The snap to nearest leg segment is only possible once the leg has been defined so that can't be used. However, I have implemented a similar approach for `getExpectedLeg` which should result in a lot more matches.